[python-package] add PyArrow Table support to get_data and add_features_from methods

suk1yak1 opened the pull request:
Fixes #6891

## Description
This PR adds proper support for PyArrow Table in Dataset's get_data and add_features_from methods. 
Previously, when using a PyArrow Table as input data, there was no handling for:
1. Subsetting a PyArrow Table in get_data method when used_indices is set
2. Merging PyArrow Tables in add_features_from method

These missing implementations caused issues when working with PyArrow Tables with free_raw_data=False, 
as the data would be incorrectly set to None after operations.

## Changes
1. Added PyArrow Table support in get_data method to handle subsetting with used_indices
2. Added PyArrow Table support in add_features_from method to properly merge columns from two tables
3. Added proper validation and error handling when working with PyArrow Tables

suk1yak1 committed:
[python-package]add_features_from with PyArrow Table incorrectly frees raw data despite free_raw_data=False (#6891)

jameslamb commented:
Thanks! We'll review more thoroughly later, but a first quick comment... please add some tests covering the new code you'd like this project to maintain.

suk1yak1 commented:
@microsoft-github-policy-service agree

suk1yak1 committed:
[python-package] add PyArrow Table case to test_add_features_from_different_sources (#6891)

suk1yak1 committed:
[python-package] fix handling and tests for PyArrow Table input in add_features_from method (#6891)

suk1yak1 committed:
Merge branch 'master' into fix/6891-pyarrow-table-add-features

suk1yak1 commented:
Thank you for your feedback! I have added a test case for the Pyarrow Table scenario to `test_add_features_from_different_sources`. I also updated the code to handle the case where `other.data` is a Pyarrow Table.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -3480,6 +3485,22 @@ def add_features_from(self, other: "Dataset") -> "Dataset":
+                            "without pyarrow installed. "
+                            "Install pyarrow and restart your session."
+                        )
+                    else:
```
Please remove this `else:`. We have a slight preference in this project for just raising exceptions, to reduce unnecessary extra indentation.

Like this:

https://github.com/microsoft/LightGBM/blob/a7253603e6b036274a3804d3e86acc32eb1a4a6a/python-package/lightgbm/basic.py#L2461-L2462

That's enforced by convention today... if there's some `ruff` rule that could enforce that (like `no-unnecessary-else` or something), I'd support adding it.

suk1yak1 replied:
Thanks for the feedback!
I created an issue to propose adding a Ruff rule for this pattern: [#6903](https://github.com/microsoft/LightGBM/issues/6903).
Following that, I also opened a pull request [#6904](https://github.com/microsoft/LightGBM/pull/6904) to enable the RET506 (superfluous-else-raise) rule and fix the related code.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -3521,6 +3580,79 @@ def add_features_from(self, other: "Dataset") -> "Dataset":
                     self.data = dt_DataTable(np.hstack((self.data.to_numpy(), other.data.values)))
                 elif isinstance(other.data, dt_DataTable):
                     self.data = dt_DataTable(np.hstack((self.data.to_numpy(), other.data.to_numpy())))
+                elif isinstance(other.data, pa_Table):
+                    if not PYARROW_INSTALLED:
+                        raise LightGBMError(
+                            "Cannot add features to pyarrow.Table type of raw data "
+                            "without pyarrow installed. "
+                            "Install pyarrow and restart your session."
+                        )
+                    else:
+                        self.data = dt_DataTable(
+                            np.hstack(
+                                (
+                                    self.data.to_numpy(),
+                                    np.column_stack(
+                                        [other.data.column(i).to_numpy() for i in range(len(other.data.column_names))]
+                                    ),
+                                )
+                            )
+                        )
+                else:
+                    self.data = None
```
Seeing more `datatable` code getting added here is making me think we should just do what `xgboost` did (https://github.com/dmlc/xgboost/pull/11070) and just fully drop support for it now.

In #6662, I'd proposed having deprecation warnings for "2-3 releases", but I'm going to put up a PR just proposing dropping this in the next release. We got a deprecation warning into 4.6.0 (#6670), which was released about 2 months ago, and it'll probably be at least another 2 months until the next LightGBM release... I think that's enough time.

jameslamb replied:
Opened #6894

suk1yak1 replied:
Thanks for thinking through this and for moving things forward!
I appreciate you taking the initiative to propose a clear path for dropping `datatable` support.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -3293,6 +3296,8 @@ def get_data(self) -> Optional[_LGBM_TrainDataType]:
                 elif isinstance(self.data, Sequence):
                     self.data = self.data[self.used_indices]
+                elif isinstance(self.data, pa_Table):
+                    self.data = self.data.take(self.used_indices)
```
Thanks! This was definitely just something we'd missed.

Can you please add a test in https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_arrow.py just for this change to `get_data()`? The other changes you made in `test_basic.py` do not cover these changes.

When you do that, please check that the content of `self.data` AND the returned value are correct (e.g., contain exactly the expected values and data types).

If you'd like, I'd even support opening a new pull request that only has the `get_data()` changes + test (and then making this PR only about `add_features_from()`). Totally your choice, I want to be respectful of your time.

suk1yak1 replied:
Thanks for the suggestion!
I added tests for both get_data() and add_features_from() directly in `test_arrow.py` as part of this PR.
Please let me know if there’s anything else you’d like me to adjust!

suk1yak1 replied:
@jameslamb 
I’ve opened a new pull request(https://github.com/microsoft/LightGBM/pull/6911) that includes only the changes to get_data() along with the corresponding test. This should help keep things focused. I’d appreciate it if you could take a look when you have time.

jameslamb replied:
Thanks! I'll focus there.

jameslamb reviewed (changes requested):
Thanks very much! I was able to review a bit more closely tonight, left another round of suggestions. I will put up a PR trying to drop support for h2o's `datatable`, to hopefully make this PR a little bit simpler.

suk1yak1 committed:
delete unnecessary-else

suk1yak1 committed:
Merge branch 'master' of https://github.com/microsoft/LightGBM into fix/6891-pyarrow-table-add-features

suk1yak1 committed:
[python-package]add test for pyarrow table in Dataset.get_data()

suk1yak1 committed:
[python-package]add test for add_features_from with pyarrow tables

suk1yak1 committed:
Merge branch 'master' into fix/6891-pyarrow-table-add-features

suk1yak1 committed:
[python-package] add PyArrow Table to get_data

suk1yak1 commented:
I’ll mark this PR as a draft for now. Once https://github.com/microsoft/LightGBM/pull/6911 is merged, I’ll reopen it and follow up accordingly.

suk1yak1 committed:
[python-package] add test for subset of PyArrow table dataset

StrikerRUS committed:
Merge branch 'master' into fix/6891-pyarrow-table-get-data

suk1yak1 committed:
Merge branch 'master' into fix/6891-pyarrow-table-get-data

suk1yak1 committed:
[python-package] improve PyArrow table subset tests for null values and chunking

jameslamb committed:
Merge branch 'master' into fix/6891-pyarrow-table-get-data

suk1yak1 committed:
Merge branch 'master' into fix/6891-pyarrow-table-get-data

suk1yak1 committed:
Merge branch 'master' into fix/6891-pyarrow-table-get-data

suk1yak1 committed:
[python-package]avoid TypeError: ChunkedArray.to_numpy() takes no keyword arguments

suk1yak1 committed:
Merge branch 'master' into fix/6891-pyarrow-table-get-data

suk1yak1 committed:
[python-package]Move final assertion before for loop to fail faster and group related comparisons

Co-authored-by: James Lamb <jaylamb20@gmail.com>

suk1yak1 committed:
[python-package]Rename test helper and add docstring to clarify purpose

Co-authored-by: James Lamb <jaylamb20@gmail.com>

suk1yak1 committed:
[python-package]Rename test helper

suk1yak1 committed:
Merge branch 'fix/6891-pyarrow-table-get-data' of github.com:suk1yak1/LightGBM into fix/6891-pyarrow-table-add-features
